File: PeterO/Rdf/RDFTerm.cs
/*
Written in 2013 by Peter Occil.
Any copyright is dedicated to the Public Domain.
http://creativecommons.org/publicdomain/zero/1.0/

If you like this, you should donate to Peter O.
at: http://upokecenter.dreamhosters.com/articles/donate-now-2/
*/
namespace PeterO.Rdf {
using System;
using System.Text;

    /// <summary>Not documented yet.</summary>
public sealed class RDFTerm {
    /// <summary>Type value for a blank node.</summary>
  public const int BLANK = 0;  // type is blank node name, literal is blank

    /// <summary>Type value for an IRI (Internationalized Resource
    /// Identifier.).</summary>
  public const int IRI = 1;  // type is IRI, literal is blank

    /// <summary>Type value for a string with a language tag.</summary>
  public const int LANGSTRING = 2;  // literal is given

    /// <summary>Type value for a piece of data serialized to a
    /// _string.</summary>
  public const int TYPEDSTRING = 3;  // type is IRI, literal is given

  private static void escapeBlankNode(string str, StringBuilder builder) {
    int length = str.Length;
    string hex = "0123456789ABCDEF";
    for (int i = 0; i < length; ++i) {
      int c = str[i];
      if ((c >= 'A' && c <= 'Z') || (c>= 'a' && c<= 'z') ||
          (c > 0 && c >= '0' && c<= '9')) {
        builder.Append((char)c);
  } else if ((c & 0xfc00) == 0xd800 && i + 1 < length &&
          str[i + 1] >= 0xdc00 && str[i + 1] <= 0xdfff) {
        // Get the Unicode code point for the surrogate pair
        c = 0x10000 + (c - 0xd800)*0x400+(str[i + 1]-0xdc00);
        builder.Append("U00");
        builder.Append(hex[(c >> 20) & 15]);
        builder.Append(hex[(c >> 16) & 15]);
        builder.Append(hex[(c >> 12) & 15]);
        builder.Append(hex[(c >> 8) & 15]);
        builder.Append(hex[(c >> 4) & 15]);
        builder.Append(hex[c & 15]);
        ++i;
      } else {
        builder.Append("u");
        builder.Append(hex[(c >> 12) & 15]);
        builder.Append(hex[(c >> 8) & 15]);
        builder.Append(hex[(c >> 4) & 15]);
        builder.Append(hex[c & 15]);
      }
    }
  }

  private static void escapeLanguageTag(string str, StringBuilder builder) {
    int length = str.Length;
    var hyphen = false;
    for (int i = 0; i < length; ++i) {
      int c = str[i];
      if (c >= 'A' && c <= 'Z') {
        builder.Append((char)(c + 0x20));
      } else if (c >= 'a' && c <= 'z') {
        builder.Append((char)c);
      } else if (hyphen && c >= '0' && c <= '9') {
        builder.Append((char)c);
      } else if (c == '-') {
        builder.Append((char)c);
        hyphen = true;
        if (i + 1 < length && str[i+1]=='-') {
          builder.Append('x');
        }
      } else {
        builder.Append('x');
      }
    }
  }

  private static void escapeString(
string str,
StringBuilder builder,
bool uri) {
    int length = str.Length;
    string hex = "0123456789ABCDEF";
    for (int i = 0; i < length; ++i) {
      int c = str[i];
      if (c == 0x09) {
        builder.Append("\\t");
      } else if (c == 0x0a) {
        builder.Append("\\n");
      } else if (c == 0x0d) {
        builder.Append("\\r");
      } else if (c == 0x22) {
        builder.Append("\\\"");
      } else if (c == 0x5c) {
        builder.Append("\\\\");
      } else if (uri && c == '>') {
        builder.Append("%3E");
      } else if (c >= 0x20 && c <= 0x7e) {
        builder.Append((char)c);
  } else if ((c & 0xfc00) == 0xd800 && i + 1 < length &&
          str[i + 1] >= 0xdc00 && str[i + 1] <= 0xdfff) {
        // Get the Unicode code point for the surrogate pair
        c = 0x10000 + (c - 0xd800)*0x400+(str[i + 1]-0xdc00);
        builder.Append("\\U00");
        builder.Append(hex[(c >> 20) & 15]);
        builder.Append(hex[(c >> 16) & 15]);
        builder.Append(hex[(c >> 12) & 15]);
        builder.Append(hex[(c >> 8) & 15]);
        builder.Append(hex[(c >> 4) & 15]);
        builder.Append(hex[c & 15]);
        ++i;
      } else {
        builder.Append("\\u");
        builder.Append(hex[(c >> 12) & 15]);
        builder.Append(hex[(c >> 8) & 15]);
        builder.Append(hex[(c >> 4) & 15]);
        builder.Append(hex[c & 15]);
      }
    }
  }

  private string typeOrLanguage = null;
  private string value = null;
  private int kind;

    /// <summary>Predicate for RDF types.</summary>
  public static readonly RDFTerm A =
      fromIRI("http://www.w3.org/1999/02/22-rdf-syntax-ns#type");

    /// <summary>Predicate for the first object in a list.</summary>
  public static readonly RDFTerm FIRST = fromIRI(
      "http://www.w3.org/1999/02/22-rdf-syntax-ns#first");

    /// <summary>Object for nil, the end of a list, or an empty
    /// list.</summary>
  public static readonly RDFTerm NIL = fromIRI(
      "http://www.w3.org/1999/02/22-rdf-syntax-ns#nil");

    /// <summary>Predicate for the remaining objects in a list.</summary>
  public static readonly RDFTerm REST = fromIRI(
      "http://www.w3.org/1999/02/22-rdf-syntax-ns#rest");

    /// <summary>Object for false.</summary>
  public static readonly RDFTerm FALSE = fromTypedString(
      "false",
      "http://www.w3.org/2001/XMLSchema#bool");

    /// <summary>Object for true.</summary>
  public static readonly RDFTerm TRUE = fromTypedString(
      "true",
      "http://www.w3.org/2001/XMLSchema#bool");

    /// <summary>Not documented yet.</summary>
    /// <param name='name'>Not documented yet.</param>
    /// <returns>A RDFTerm object.</returns>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='name'/> is null.</exception>
  public static RDFTerm fromBlankNode(string name) {
    if (name == null) {
 throw new ArgumentNullException("name");
}
    if (name.Length == 0) {
 throw new ArgumentException("name is empty.");
}
    var ret = new RDFTerm();
    ret.kind = BLANK;
    ret.typeOrLanguage = null;
    ret.value = name;
    return ret;
  }

    /// <summary>Not documented yet.</summary>
    /// <param name='iri'>Not documented yet.</param>
    /// <returns>A RDFTerm object.</returns>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='iri'/> is null.</exception>
  public static RDFTerm fromIRI(string iri) {
    if (iri == null) {
 throw new ArgumentNullException("iri");
}
    var ret = new RDFTerm();
    ret.kind = IRI;
    ret.typeOrLanguage = null;
    ret.value = iri;
    return ret;
  }

    /// <summary>Not documented yet.</summary>
    /// <param name='str'>Not documented yet.</param>
    /// <param name='languageTag'>Not documented yet.</param>
    /// <returns>A RDFTerm object.</returns>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='str'/> or <paramref name='languageTag'/> is null.</exception>
  public static RDFTerm fromLangString(string str, string languageTag) {
    if (str == null) {
 throw new ArgumentNullException("str");
}
    if (languageTag == null) {
 throw new ArgumentNullException("languageTag");
}
    if (languageTag.Length == 0) {
 throw new ArgumentException("languageTag is empty.");
}
    var ret = new RDFTerm();
    ret.kind = LANGSTRING;
    ret.typeOrLanguage = languageTag;
    ret.value = str;
    return ret;
  }

    /// <summary>Not documented yet.</summary>
    /// <param name='str'>Not documented yet.</param>
    /// <returns>A RDFTerm object.</returns>
  public static RDFTerm fromTypedString(string str) {
    return fromTypedString(str, "http://www.w3.org/2001/XMLSchema#string");
  }

    /// <summary>Not documented yet.</summary>
    /// <param name='str'>Not documented yet.</param>
    /// <param name='iri'>Not documented yet.</param>
    /// <returns>A RDFTerm object.</returns>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='str'/> or <paramref name='iri'/> is null.</exception>
  public static RDFTerm fromTypedString(string str, string iri) {
    if (str == null) {
 throw new ArgumentNullException("str");
}
    if (iri == null) {
 throw new ArgumentNullException("iri");
}
    if (iri.Length == 0) {
 throw new ArgumentException("iri is empty.");
}
    var ret = new RDFTerm();
    ret.kind = TYPEDSTRING;
    ret.typeOrLanguage = iri;
    ret.value = str;
    return ret;
  }

    /// <summary>Not documented yet.</summary>
    /// <returns>Not documented yet.</returns>
  public override sealed bool Equals(object obj) {
    if (this == obj) {
 return true;
}
    if (obj == null) {
 return false;
}
    if (GetType() != obj.GetType()) {
 return false;
}
    var other = (RDFTerm) obj;
    if (this.kind != other.kind) {
 return false;
}
    if (this.typeOrLanguage == null) {
      if (other.typeOrLanguage != null) {
 return false;
}
    } else if (!this.typeOrLanguage.Equals(other.typeOrLanguage)) {
 return false;
}
    if (this.value == null) {
      return other.value != null;
    } else {
 return !this.value.Equals(other.value);
}
  }

    /// <summary>Not documented yet.</summary>
    /// <returns>A 32-bit signed integer.</returns>
  public int getKind() {
    return this.kind;
  }

    /// <summary>Gets the language tag or data type for this RDF
    /// literal.</summary>
    /// <returns>A string object.</returns>
  public string getTypeOrLanguage() {
    return this.typeOrLanguage;
  }

    /// <summary>Gets the IRI, blank node identifier, or lexical form of an
    /// RDF literal.</summary>
    /// <returns>A string object.</returns>
  public string getValue() {
    return this.value;
  }

    /// <summary>Not documented yet.</summary>
    /// <returns>Not documented yet.</returns>
  public override sealed int GetHashCode() {unchecked {
     var prime = 31;
    int result = prime + this.kind;
    result = (prime * result) + ((this.typeOrLanguage == null) ? 0 :
            this.typeOrLanguage.GetHashCode());
    result = prime * result + ((this.value = = null) ? 0 :
      this.value.GetHashCode());
    return result;
  }}

    /// <summary>Not documented yet.</summary>
    /// <returns>A Boolean object.</returns>
  public bool isBlank() {
    return this.kind == BLANK;
  }

    /// <summary>Not documented yet.</summary>
    /// <param name='str'>Not documented yet.</param>
    /// <returns>A Boolean object.</returns>
  public bool isIRI(string str) {
    return this.kind == IRI && str != null && str.Equals(this.value);
  }

    /// <summary>Not documented yet.</summary>
    /// <returns>A Boolean object.</returns>
  public bool isOrdinaryString() {
    return this.kind == TYPEDSTRING && "http://www.w3.org/2001/XMLSchema#string"
      .Equals(this.typeOrLanguage);
  }

    /// <summary>* Gets a _string representation of this RDF term in
    /// N-Triples format. The _string will not end in a line
    /// break.</summary>
    /// <returns>Not documented yet.</returns>
  public override sealed string ToString() {
    StringBuilder builder = null;
    if (this.kind == BLANK) {
      builder = new StringBuilder();
      builder.Append("_:");
      escapeBlankNode(this.value, builder);
    } else if (this.kind == LANGSTRING) {
      builder = new StringBuilder();
      builder.Append("\"");
      escapeString(this.value, builder, false);
      builder.Append("\"@");
      escapeLanguageTag(this.typeOrLanguage, builder);
    } else if (this.kind == TYPEDSTRING) {
      builder = new StringBuilder();
      builder.Append("\"");
      escapeString(this.value, builder, false);
      builder.Append("\"");
  if (!"http://www.w3.org/2001/XMLSchema#string"
        .Equals(this.typeOrLanguage)) {
        builder.Append("^^<");
        escapeString(this.typeOrLanguage, builder, true);
        builder.Append(">");
      }
    } else if (this.kind == IRI) {
      builder = new StringBuilder();
      builder.Append("<");
      escapeString(this.value, builder, true);
      builder.Append(">");
    } else {
 return "<about:blank>";
}
    return builder.ToString();
  }
}
}

File: src/main/java/com/upokecenter/rdf/RDFTerm.java
package com.upokecenter.util;
/*
Written in 2013 by Peter Occil.
Any copyright is dedicated to the Public Domain.
http://creativecommons.org/publicdomain/zero/1.0/

If you like this, you should donate to Peter O.
at: http://peteroupc.github.io/
*/

    /**
     * Not documented yet.
     */
  public final class RDFTerm {
    /**
     * Type value for a blank node.
     */
    public static final int BLANK = 0;  // type is blank node name, literal is blank

    /**
     * Type value for an IRI (Internationalized Resource Identifier.).
     */
    public static final int IRI = 1;  // type is IRI, literal is blank

    /**
     * Type value for a string with a language tag.
     */
    public static final int LANGSTRING = 2;  // literal is given

    /**
     * Type value for a piece of data serialized to a string.
     */
    public static final int TYPEDSTRING = 3;  // type is IRI, literal is given

    private static void escapeBlankNode(String str, StringBuilder builder) {
      int length = str.length();
      String hex = "0123456789ABCDEF";
      for (int i = 0; i < length; ++i) {
        int c = str.charAt(i);
        if ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') ||
            (c > 0 && c >= '0' && c <= '9')) {
          builder.append((char)c);
        } else if ((c & 0xfc00) == 0xd800 && i + 1 < length &&
                str.charAt(i + 1) >= 0xdc00 && str.charAt(i + 1) <= 0xdfff) {
          // Get the Unicode code point for the surrogate pair
          c = 0x10000 + ((c - 0xd800) << 10) + (str.charAt(i + 1) - 0xdc00);
          builder.append("U00");
          builder.append(hex.charAt((c >> 20) & 15));
          builder.append(hex.charAt((c >> 16) & 15));
          builder.append(hex.charAt((c >> 12) & 15));
          builder.append(hex.charAt((c >> 8) & 15));
          builder.append(hex.charAt((c >> 4) & 15));
          builder.append(hex.charAt(c & 15));
          ++i;
        } else {
          builder.append("u");
          builder.append(hex.charAt((c >> 12) & 15));
          builder.append(hex.charAt((c >> 8) & 15));
          builder.append(hex.charAt((c >> 4) & 15));
          builder.append(hex.charAt(c & 15));
        }
      }
    }

    private static void escapeLanguageTag(String str, StringBuilder builder) {
      int length = str.length();
      boolean hyphen = false;
      for (int i = 0; i < length; ++i) {
        int c = str.charAt(i);
        if (c >= 'A' && c <= 'Z') {
          builder.append((char)(c + 0x20));
        } else if (c >= 'a' && c <= 'z') {
          builder.append((char)c);
        } else if (hyphen && c >= '0' && c <= '9') {
          builder.append((char)c);
        } else if (c == '-') {
          builder.append((char)c);
          hyphen = true;
          if (i + 1 < length && str.charAt(i + 1) == '-') {
            builder.append('x');
          }
        } else {
          builder.append('x');
        }
      }
    }

    private static void escapeString(
  String str,
  StringBuilder builder,
  boolean uri) {
      int length = str.length();
      String Hex = "0123456789ABCDEF";
      for (int i = 0; i < length; ++i) {
        int c = str.charAt(i);
        if (c == 0x09) {
          builder.append("\\t");
        } else if (c == 0x0a) {
          builder.append("\\n");
        } else if (c == 0x0d) {
          builder.append("\\r");
        } else if (c == 0x22) {
          builder.append("\\\"");
        } else if (c == 0x5c) {
          builder.append("\\\\");
        } else if (uri && c == '>') {
          builder.append("%3E");
        } else if (c >= 0x20 && c <= 0x7e) {
          builder.append((char)c);
        } else if ((c & 0xfc00) == 0xd800 && i + 1 < length &&
                str.charAt(i + 1) >= 0xdc00 && str.charAt(i + 1) <= 0xdfff) {
          // Get the Unicode code point for the surrogate pair
          c = 0x10000 + ((c - 0xd800) << 10) + (str.charAt(i + 1) - 0xdc00);
          builder.append("\\U00");
          builder.append(Hex.charAt((c >> 20) & 15));
          builder.append(Hex.charAt((c >> 16) & 15));
          builder.append(Hex.charAt((c >> 12) & 15));
          builder.append(Hex.charAt((c >> 8) & 15));
          builder.append(Hex.charAt((c >> 4) & 15));
          builder.append(Hex.charAt(c & 15));
          ++i;
        } else {
          builder.append("\\u");
          builder.append(Hex.charAt((c >> 12) & 15));
          builder.append(Hex.charAt((c >> 8) & 15));
          builder.append(Hex.charAt((c >> 4) & 15));
          builder.append(Hex.charAt(c & 15));
        }
      }
    }

    private String typeOrLanguage = null;
    private String value = null;
    private int kind;

    /**
     * Predicate for RDF types.
     */
    public static final RDFTerm A =
        fromIRI("http://www.w3.org/1999/02/22-rdf-syntax-ns#type");

    /**
     * Predicate for the first object in a list.
     */
    public static final RDFTerm FIRST = fromIRI(
        "http://www.w3.org/1999/02/22-rdf-syntax-ns#first");

    /**
     * Object for nil, the end of a list, or an empty list.
     */
    public static final RDFTerm NIL = fromIRI(
        "http://www.w3.org/1999/02/22-rdf-syntax-ns#nil");

    /**
     * Predicate for the remaining objects in a list.
     */
    public static final RDFTerm REST = fromIRI(
        "http://www.w3.org/1999/02/22-rdf-syntax-ns#rest");

    /**
     * Object for false.
     */
    public static final RDFTerm FALSE = fromTypedString(
        "false",
        "http://www.w3.org/2001/XMLSchema#boolean");

    /**
     * Object for true.
     */
    public static final RDFTerm TRUE = fromTypedString(
        "true",
        "http://www.w3.org/2001/XMLSchema#boolean");

    /**
     * Not documented yet.
     * @param name The parameter {@code name} is not documented yet.
     * @return A RDFTerm object.
     * @throws java.lang.NullPointerException The parameter {@code name} is null.
     */
    public static RDFTerm fromBlankNode(String name) {
      if (name == null) {
        throw new NullPointerException("name");
      }
      if (name.length() == 0) {
        throw new IllegalArgumentException("name is empty.");
      }
      RDFTerm ret = new RDFTerm();
      ret.kind = BLANK;
      ret.typeOrLanguage = null;
      ret.value = name;
      return ret;
    }

    /**
     * Not documented yet.
     * @param iri The parameter {@code iri} is not documented yet.
     * @return A RDFTerm object.
     * @throws java.lang.NullPointerException The parameter {@code iri} is null.
     */
    public static RDFTerm fromIRI(String iri) {
      if (iri == null) {
        throw new NullPointerException("iri");
      }
      RDFTerm ret = new RDFTerm();
      ret.kind = IRI;
      ret.typeOrLanguage = null;
      ret.value = iri;
      return ret;
    }

    /**
     * Not documented yet.
     * @param str The parameter {@code str} is not documented yet.
     * @param languageTag The parameter {@code languageTag} is not documented yet.
     * @return A RDFTerm object.
     * @throws java.lang.NullPointerException The parameter {@code str} or {@code
     * languageTag} is null.
     */
    public static RDFTerm fromLangString(String str, String languageTag) {
      if (str == null) {
        throw new NullPointerException("str");
      }
      if (languageTag == null) {
        throw new NullPointerException("languageTag");
      }
      if (languageTag.length() == 0) {
        throw new IllegalArgumentException("languageTag is empty.");
      }
      RDFTerm ret = new RDFTerm();
      ret.kind = LANGSTRING;
      ret.typeOrLanguage = languageTag;
      ret.value = str;
      return ret;
    }

    /**
     * Not documented yet.
     * @param str The parameter {@code str} is not documented yet.
     * @return A RDFTerm object.
     */
    public static RDFTerm fromTypedString(String str) {
      return fromTypedString(str, "http://www.w3.org/2001/XMLSchema#String");
    }

    /**
     * Not documented yet.
     * @param str The parameter {@code str} is not documented yet.
     * @param iri The parameter {@code iri} is not documented yet.
     * @return A RDFTerm object.
     * @throws java.lang.NullPointerException The parameter {@code str} or {@code
     * iri} is null.
     */
    public static RDFTerm fromTypedString(String str, String iri) {
      if (str == null) {
        throw new NullPointerException("str");
      }
      if (iri == null) {
        throw new NullPointerException("iri");
      }
      if (iri.length() == 0) {
        throw new IllegalArgumentException("iri is empty.");
      }
      RDFTerm ret = new RDFTerm();
      ret.kind = TYPEDSTRING;
      ret.typeOrLanguage = iri;
      ret.value = str;
      return ret;
    }

    /**
     * Not documented yet.
     * @return The return value is not documented yet.
     */
    @Override public final boolean equals(Object obj) {
      if (this == obj) {
        return true;
      }
      if (obj == null) {
        return false;
      }
      if (GetType() != obj.getClass()) {
        return false;
      }
      RDFTerm other = (RDFTerm)obj;
      if (this.kind != other.kind) {
        return false;
      }
      if (this.typeOrLanguage == null) {
        if (other.typeOrLanguage != null) {
          return false;
        }
      } else if (!this.typeOrLanguage.equals(other.typeOrLanguage)) {
        return false;
      }
      if (this.value == null) {
        return other.value != null;
      } else {
        return !this.value.equals(other.value);
      }
    }

    /**
     * Not documented yet.
     * @return A 32-bit signed integer.
     */
    public int getKind() {
      return this.kind;
    }

    /**
     * Gets the language tag or data type for this RDF literal.
     * @return A text string.
     */
    public String getTypeOrLanguage() {
      return this.typeOrLanguage;
    }

    /**
     * Gets the IRI, blank node identifier, or lexical form of an RDF literal.
     * @return A text string.
     */
    public String getValue() {
      return this.value;
    }

    /**
     * Not documented yet.
     * @return The return value is not documented yet.
     */
    @Override public final int hashCode() {
      {
        int prime = 31;
        int result = prime + this.kind;
        result = (prime * result) + ((this.typeOrLanguage == null) ? 0 :
                this.typeOrLanguage.hashCode());
        boolean isnull = this.value == null;
        result = (prime * result) + (isnull ? 0 : this.value.hashCode());
        return result;
      }
    }

    /**
     * Gets whether this term is a blank node.
     * @return A Boolean object.
     */
    public boolean isBlank() {
      return this.kind == BLANK;
    }

    /**
     * Not documented yet.
     * @param str The parameter {@code str} is not documented yet.
     * @return A Boolean object.
     */
    public boolean isIRI(String str) {
      return this.kind == IRI && str != null && str.equals(this.value);
    }

    /**
     * Not documented yet.
     * @return A Boolean object.
     */
    public boolean isOrdinaryString() {
   return this.kind == TYPEDSTRING && "http://www.w3.org/2001/XMLSchema#String"
        .equals(this.typeOrLanguage);
    }

    /**
     * Gets a string representation of this RDF term in N-Triples format. The
     * string will not end in a line break.
     * @return The return value is not documented yet.
     */
    @Override public final String toString() {
      StringBuilder builder = null;
      if (this.kind == BLANK) {
        builder = new StringBuilder();
        builder.append("_:");
        escapeBlankNode(this.value, builder);
      } else if (this.kind == LANGSTRING) {
        builder = new StringBuilder();
        builder.append("\"");
        escapeString(this.value, builder, false);
        builder.append("\"@");
        escapeLanguageTag(this.typeOrLanguage, builder);
      } else if (this.kind == TYPEDSTRING) {
        builder = new StringBuilder();
        builder.append("\"");
        escapeString(this.value, builder, false);
        builder.append("\"");
        if (!"http://www.w3.org/2001/XMLSchema#String"
              .equals(this.typeOrLanguage)) {
          builder.append("^^<");
          escapeString(this.typeOrLanguage, builder, true);
          builder.append(">");
        }
      } else if (this.kind == IRI) {
        builder = new StringBuilder();
        builder.append("<");
        escapeString(this.value, builder, true);
        builder.append(">");
      } else {
        return "<>";
      }
      return builder.toString();
    }
  }
